File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
def solution(orders, course):
    answer = []
    d = {}
    for i in course :
        d[i] = {}
    for i in orders :
        t = list(i)
        t.sort()
        for j in course :
            if len(t) < j : continue
            for k in combinations(t, j) :
                if k in d[j] : d[j][k] += 1
                else : d[j][k] = 1
    
    temp = {}
    for i in course :
        temp[i] = []
        for j in d[i] :
            if d[i][j] >= 2 :
                temp[i].append([j, d[i][j]])
    
    for i in temp :
        temp[i].sort(key = lambda x : (len(x[0]), -x[1]))
    
    for i in course :
        for j in temp :
            if i == j :
                for index, k in enumerate(temp[j]) :
                    if index == 0 :
                        m = k[1]
                    if m == k[1] :
                        answer.append(''.join(k[0]))
                        
    answer.sort()
    return answer

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    public static String[] solution(String[] orders, int[] course) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < orders.length; i++) {
            for (int j = 0; j < course.length; j++) {
                if (orders[i].length() < course[j]) {
                    break;
                }
                String[] target = orders[i].split("");
                Arrays.sort(target);
                int length = course[j];

                dfs(target, length, 0, new ArrayList<>(), map);
            }
        }

        List<String> answer = new ArrayList<>();
        for (int i = 0; i < course.length; i++) {
            List<Map.Entry<String, Integer>> temp = new ArrayList<>();
            for (Map.Entry<String, Integer> entry : map.entrySet()) {
                if (entry.getKey().length() == course[i] && entry.getValue() >= 2) {
                    temp.add(entry);
                }
            }

            int max = temp.stream()
                    .mapToInt(entry -> entry.getValue())
                    .max()
                    .orElse(0);

            for (Map.Entry<String, Integer> entry : temp) {
                if (entry.getValue() == max) {
                    answer.add(entry.getKey());
                }
            }
        }

        Collections.sort(answer);
        return answer.toArray(String[]::new);
    }

    public static void dfs(String[] target, int r, int start,
                           List<String> current, Map<String, Integer> result) {
        if (current.size() == r) {
            result.put(String.join("", current),
                    result.getOrDefault(String.join("", current), 0) + 1);
            return;
        }

        if (target.length - start < r - current.size()) {
            return;
        }

        for (int i = start; i < target.length; i++) {
            current.add(target[i]);
            dfs(target, r, i + 1, current, result);
            current.remove(current.size() - 1);
        }

    }
}